server/src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -873,7 +883,27 @@ public CacheHelper getReaderCacheHelper() {
             super(in, new FilterDirectoryReader.SubReaderWrapper() {
                 @Override
                 public LeafReader wrap(LeafReader leaf) {
-                    return new SubReaderWithAllLiveDocs(leaf);
+                    try {
+                        if (leaf.getLiveDocs() == null) {
+                            return leaf;
+                        }
+                        DocIdSetIterator rollbackDocs = DocValuesFieldExistsQuery.getDocValuesDocIdSetIterator(ROLLED_BACK_FIELD, leaf);
+                        if (rollbackDocs == null) {
+                            return new SubReaderWithLiveDocs(leaf, null, leaf.maxDoc());
+                        }
+                        int rollbackCount = 0;
+                        FixedBitSet liveDocs = new FixedBitSet(leaf.maxDoc());
```

Comment:
@jpountz I see @s1monw making a reader cache for `SoftDeletesDirectoryReaderWrapper`. However, I could not create a instance of `CacheKey` because its constructor is not public.

Reply:
I am exploring a cache of liveDocs using coreCacheHelper and docValuesGeneration as key. It looks good, but I need to make sure that it works correctly and write some tests.

Reply:
I think you don't need to create cache keys and could directly use LeafReader instances as cache keys.